### ISSUE TITLE
sub/section number in references is wrong, when using AtBegin

### ISSUE THREAD

**Issue by josephwright** [josephwright/beamer#118]
hi

in the attached example, the references to labels of sections and subsections are wrong.

The problem is that the LaTeX `\section{sec3}\label{sec3}` works by writing the content of `\@currentlabel` into the `test.aux` file. But if
````
\AtBeginSection[]{
\begin{frame}<presentation>[plain]
\frametitle{Outline of section}
\tableofcontents[currentsection] 
end{frame}}
````
is used, the 'frame' environment damages the value of `\@currentlabel`

The 'frame' environment should store the value of `\@currentlabe`l somewhere and put it back when it ends. (Thats what all other LaTeX environment such as theorems and equations do).

I attach a test case, and a a patch.

a.

ps: this was posted as bug 2089481 in sourceforge, together with a patch, but the patch was never applied; I hope this time it is.
````
\documentclass{beamer}

\usetheme[secheader]{Boadilla}

%if you comment out the following AtBegin*, labeling works
 \AtBeginSection[]
 {
   \begin{frame}<presentation>[plain]
     \frametitle{Outline of section}
     \tableofcontents[currentsection]
   \end{frame}
 }


 \AtBeginSubsection[]
 {
   \begin{frame}<presentation>[plain]
     \frametitle{Outline of subsection}
     \tableofcontents[currentsection,currentsubsection]
   \end{frame}
 }



\begin{document}

\section{sec1}

\begin{frame}
  the following labels are  all wrong

  see section \ref{sec5} (should be section 5)

  see section \ref{sec3} (should be section 3)

  see subsection \ref{sec5sub2} (should be 5.2, or maybe just 2)
  \end{frame}
\subsection{sec1sub1}
\begin{frame}
  frame  \insertframenumber
\end{frame}
\section{sec2}
\begin{frame}
  frame  \insertframenumber
\end{frame}
\section{sec3}\label{sec3}
\begin{frame}
  frame  \insertframenumber
\end{frame}
\section{sec4}
\begin{frame}
  frame  \insertframenumber
\end{frame}
\subsection{sec4sub1}\label{sec4sub1}
\begin{frame}
  frame  \insertframenumber
\end{frame}
\section{sec5}\label{sec5}

\begin{frame}
  frame  \insertframenumber
\end{frame}
\begin{frame}
  frame  \insertframenumber
  \pause
  bis
\end{frame}
\subsection{sec5sub1}\label{sec5sub1}
\begin{frame}
  frame  \insertframenumber
\end{frame}
\subsection{sec5sub2}\label{sec5sub2}
\begin{frame}
  frame  \insertframenumber
\end{frame}
\end{document}
````
Patch
````
--- beamer/base/beamerbaseframe.sty	2010-06-17 11:19:15.000000000 +0200
+++ /usr/share/texmf/tex/latex/beamer/base/beamerbaseframe.sty	2011-02-25 19:50:11.058387470 +0100
@@ -27,6 +27,9 @@
   \beamer@inframefalse
 }
 
+\def\refcounter#1{\protected@edef\@currentlabel%
+       {\csname p@#1\endcsname\csname the#1\endcsname}}%
+
 
 %
 % Slide Definitions
@@ -79,6 +82,7 @@
     \beamer@inframetrue%
     \let\frame=\framelatex% inside frames, use LaTeX's \frame command
     \begin{beamer@framepauses}%
+      \refcounter{framenumber}%use frame number for  \label
       \ifbeamer@shrink%
         \hsize=\beamer@shrinkfactorinv\hsize%
         \textwidth=\beamer@shrinkfactorinv\textwidth%
@@ -435,7 +439,7 @@
   \ifbeamer@noframenumbering%
     \beamer@noframenumberingfalse%
   \else%
-    \refstepcounter{framenumber}%
+    \stepcounter{framenumber}%do not set \label here - it interferes with AtBegin*
   \fi%
   {\let\@elt\beamer@savecounter\beamer@overlaycounterresets}
   \gdef\beamer@framestartpage{}%
````
----------------------------------------
- Bitbucket: https://bitbucket.org/rivanvx/beamer/issue/118
- Originally reported by: A Mennucci
- Originally created at:  2011-02-26

**Comment by josephwright** [comment]
_From A Mennucci on 2011-02-26_

BTW: when I originally wrote the patches (for sourceforge), I solved the issues 117 and 118 alltogether; in presenting them here, I split it in two; but it may be the case that I got it wrong, so that you have to use both patches together to fix the issues. That is, I can guarantee that if you apply both patches in #117 and #118 , then the issues are fixed (I tested it); but I am not so sure that I splitted them right). a.

**Comment by josephwright** [comment]
Fixed by 27cc045c3ac92db07156ee1aa987b3ad6b62c916.
